system/jlib/jstats.cpp

```diff
@@ -4011,6 +4051,277 @@ void verifyStatisticFunctions()
     }
 }
 
+class StatsAggregator : public CInterface
+{
+    class ScopeId: public StatsScopeId
+    {
+    public:
+        ScopeId(StatisticScopeType _scopeType=SSTglobal, unsigned _id=0) : StatsScopeId(_scopeType, _id) {}
+        ScopeId(const ScopeId & _id) : StatsScopeId(_id.scopeType, _id.id, _id.extra) {}
+        ScopeId(const StatsScopeId & other) : StatsScopeId(other) {}
+        size_t operator()(const ScopeId & _scopeId) const
+        {
+            return std::hash<unsigned>{}((unsigned)(_scopeId.id|(_scopeId.scopeType<<16)));
+        }
+        bool operator()(const ScopeId & lhs, const ScopeId & rhs) const
+        {
+            return (lhs.scopeType==rhs.scopeType) && (lhs.id==rhs.id);
+        }
+    };
+    struct StatisticKindFunc  // funcs required by unordered_map<StatisticKind,..>
+    {
+        std::size_t operator()(const StatisticKind & sk) const
+        {
+            return std::hash<unsigned>{}((unsigned)sk);
+        }
+        bool operator()(const StatisticKind & lhs, const StatisticKind & rhs) const
+        {
+            return lhs==rhs;
+        }
+    };
+    ScopeId id;
+    std::unordered_map<StatisticKind, unsigned __int64, StatisticKindFunc, StatisticKindFunc> values;
+    std::unordered_map<ScopeId, Owned<StatsAggregator>, ScopeId, ScopeId> children;
+
+    StatsAggregator * queryAggregator(const ScopeId & scopeId)
+    {
+        if (auto search = children.find(scopeId); search != children.end())
+            return search->second;
+        return nullptr;
+    }
+    void eraseAggregator(const ScopeId & scopeId)
+    {
+        if (auto search = children.find(scopeId); search != children.end())
+            children.erase(search);
+    }
+public:
+    StatsAggregator() {}
+    StatsAggregator(ScopeId _id) : id(_id) {}
+    StatsAggregator * ensureScopeAggregator(const StatsScopeId & statScopeId)
+    {
+        ScopeId searchScope(statScopeId);
+        return ensureScopeAggregator(searchScope);
+    }
+    StatsAggregator * ensureScopeAggregator(const ScopeId & scopeId)
+    {
+        StatsAggregator * aggregator = queryAggregator(scopeId);
```

Comment:
Is this common?  If so more efficient/simpler to use map[scopeId]

Reply:
If map[scopeId] is referenced and it doesn't exist, it will create an entry for that item.  queryAggregator() queries the item but doesn't create an entry for the item.  This is important elsewhere (e.g. line 4142 & 4147).